Enable Link time optimisation for GCC

deece opened the pull request:
### Description
This pull request enables Link Time Optimisation for GCC.

This enables smaller final binaries, which is particularly important when writing bootloaders. (In my case, the binary was too big to fit under 16kB without this)

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

deece committed:
Enable LTO for GCC compilers

This patch enables Link Time Optimisation against libmbed,
to allow for smaller final binaries.

This is especially important when building bootloaders, which
have a much smaller flash area that they must fit into.

Signed-off-by: Alastair D'Silva <alastair@d-silva.org>

deece committed:
Use gcc-ar to handle archiving

Since the archive now contains LTO objects, we need to use the
LTO plugin to index the LTO objects correctly.

The gcc wrapped ar sets up the plugin path correctly.

Signed-off-by: Alastair D'Silva <alastair@d-silva.org>

cmonr commented:
@deece What kind of tests have you run with this option on? 
Were you aware of and been caught up on a related PR that is still open? https://github.com/ARMmbed/mbed-os/pull/7596 (It's for the ARM compiler release profile only)

Finally, please take a look at the troubled PR jobs.

deece commented:
@cmonr I've been building mbed-lib with this for quite a while (a good year or so), targetting STM32F030R8 and using the binaries it produces.

I missed that PR when I was searching (maybe I only searched for issues rather than PRs). This is a bit more comprehensive though. We disable partitioning to give the best opportunity for optimisation (since mbed applications are relatively small), and using gcc-ar is a necessity, as without it (or specifying the --plugin arg to the archiver), the LTO objects are missed from the archive index, resulting in bogus missing symbols at final link time.

I had a look at the CI failures and I can reproduce them without my patches, so I'm going to blame someone else for those :)

deece commented:
Compilers tested are ARM official gcc-arm-none-eabi-7-2018-q2-update & a locally built 8.2.0 (via crosstool-ng)

deepikabhavnani commented:
> I had a look at the CI failures and I can reproduce them without my patches, so I'm going to blame someone else for those :)

Please raise the respective issues and failures with steps to reproduce, those should be fixed before this PR is merged....

deece commented:
@deepikabhavnani Done, Issue #9082

deepikabhavnani commented:
@deece - Please note the commands in CI are 
```

$ sed -n '/``` c++/,/```/{/```/d;p;}' $LITTLEFS/README.md > main.cpp

$ python tools/make.py -t GCC_ARM -m K82F --source=. --build=BUILD/K82F/GCC_ARM -j0
```

deece commented:
Failure is reproducible with LTO, but is OK without.

I'll investigate when I get some time.

There's also a lot of warnings such as:
```
./features/storage/kvstore/tdbstore/TDBStore.cpp:61:3: warning: type 'struct master_record_data_t' violates the C++ One Definition Rule [-Wodr]
 } master_record_data_t;
   ^
./features/storage/nvstore/source/nvstore.cpp:66:3: note: a different type is defined in another translation unit
 } master_record_data_t;
```

which suggests that different .o's are including the same header, with different #defines, resulting in different struct layouts, all linked together. This is hidden without LTO as the compiler normally has no visibility to the code at link time, where the conflicting definitions come together.

deepikabhavnani commented:
@deece - You can raise the issues for warnings as well, respective teams will have a look. Also if you can fix and add PR for warning fixes also, it will be appreciated. Thanks

mikisch81 commented:
> Compilers tested are ARM official gcc-arm-none-eabi-7-2018-q2-update & a locally built 8.2.0 (via crosstool-ng)

@deece I also tried it with gcc 6.x which doesn't work, as I understand as long as version 6 of gcc is the official version we can't add this.

Also there is the issue of losing the memory-statistics breakdown of the final image.

In #7596 I only enabled it to the release profile so debug/develop profiles will not lose any debug abilities (and the memory statistics).

CC @kjbracey-arm

kjbracey commented:
We will be shifting to GCC 7 at some point, so this is still all useful stuff - it's added a bit more knowledge to what came up in the other PR.

IAR and ARMCC version updates are also in progress, but we've not really been thinking about GCC much, partly because there's been no obvious need for the upgrade, and partly because it's believed to be no big deal. This would be a good reason to prioritise doing it alongside the other 2 sooner rather than later.

cmonr commented:
@0xc0170 Whay was the `needs: review` label removed?

deece committed:
Common flags should also be passed to the linker when using LTO

Signed-off-by: Alastair D'Silva <alastair@d-silva.org>

deece commented:
Blocked by #9099

deece commented:
The undefined main() exists because when wrapping main, the wrapping function refers to __real_main instead, so the compiler eliminates main since it has no references. Adding the used attribute to the main prototype fixes the problem.

deece commented:
@mikisch81 The memory stats appear to be coming out fine:
```
| Module        |      .text |    .data |     .bss |
|---------------|------------|----------|----------|
| ../..         | 19337(+16) |  428(+0) | 6448(+0) |
| BUILD/K82F    |    784(+0) |    0(+0) |    0(+0) |
| [fill]        |    118(+0) |    4(+0) |  363(+0) |
| [lib]/c.a     |  31639(+0) | 2472(+0) |   89(+0) |
| [lib]/gcc.a   |   3112(+0) |    0(+0) |    0(+0) |
| [lib]/misc    |    204(+0) |    0(+0) |   28(+0) |
| [lib]/nosys.a |     32(+0) |    0(+0) |    0(+0) |
| Subtotals     | 55226(+16) | 2904(+0) | 6928(+0) |
Total Static RAM memory (data + bss): 9832(+0) bytes
Total Flash memory (text + data): 58130(+16) bytes
```

deece committed:
Mark main symbol as used

When LTO is in effect, in conjunction with wrapped main, the compiler
sees no references to main and eliminates it (the wrapper refers to
__real_main instead).

Adding the used attribute (for GCC) forces the compiler to retain the
symbol so the final link is successful.

Signed-off-by: Alastair D'Silva <alastair@d-silva.org>

deece commented:
@mikisch81 What was the problem you discovered with GCC 6? (the bug report you linked to in your PR points to the GCC 9.x)

mikisch81 commented:
@deece Sorry, I don't really remember exactly what went wrong when trying gcc 6.x.x, but on gcc 7.x.x it was ok.
For some reason I remember another link with the bug in 6.x.x..

deece commented:
@mikisch81 Ok, thanks, I'll give it a try when I get a chance

deece commented:
@cmonr @kjbracey-arm I've tested this with the following ARM GCC releases using my 1Wire softdevice code (https://github.com/InfernoEmbedded/onewire-softdevice) and they all build successfully.

gcc-arm-none-eabi-6-2017-q2-update
gcc-arm-none-eabi-7-2018-q2-update
gcc-arm-none-eabi-8-2018-q4-major

This is effectively a mbed-lib build, done by building mbed-os & disabling a bunch of features which are not supported on the STM32F030 (I had to do it this way as mbed-cli does not provide a way to build a custom mbed-lib (tracked here: https://github.com/ARMmbed/mbed-cli/issues/807)

cmonr commented:
@deece That's good to know, but I think we're holding this PR up until we have some sort of test that would make sure this doesn't break things in the future.

@ARMmbed/mbed-os-test-team I tried searching around Jira but didn't have much luck. Is there a ticket indicating if some sort of release build test will be made sometime soon?

deece commented:
Ah, I thought it was GCC 6 that was the blocker.

0xc0170 commented:
We will revisit this later (we have internal tickets for enabling lto and updating toolchains. We will provide PRs for all toolchains). I'll close this for now. 

Thanks @deece for bringing this up to our attention.
